lib_client/run_time.ml

```diff
@@ -0,0 +1,209 @@
+  | Queued_for of float (* elapsed time in seconds *)
+  | Running of { queued_for : float; ran_for : float }
+  | Finished of { queued_for : float; ran_for : float option }
+[@@deriving show]
+
+let queued_for = function
+  | Cached -> 0.
+  | Queued_for v -> v
+  | Running v -> v.queued_for
+  | Finished v -> v.queued_for
+
+let ran_for = function
+  | Cached -> 0.
+  | Queued_for _ -> 0.
+  | Running v -> v.ran_for
+  | Finished v -> Option.value ~default:0. v.ran_for
+
+let eq_timestamps (st1 : timestamps) (st2 : timestamps) =
+  match (st1, st2) with
+  | Queued v1, Queued v2 -> cmp_floats v1 v2
+  | Running v1, Running v2 ->
+      cmp_floats v1.queued_at v2.queued_at
+      && cmp_floats v1.started_at v2.started_at
+  | ( Finished
+        {
+          queued_at = queued_at1;
+          started_at = None;
+          finished_at = finished_at1;
+        },
+      Finished
+        {
+          queued_at = queued_at2;
+          started_at = None;
+          finished_at = finished_at2;
+        } ) ->
+      cmp_floats queued_at1 queued_at2 && cmp_floats finished_at1 finished_at2
+  | ( Finished
+        {
+          queued_at = queued_at1;
+          started_at = Some started_at1;
+          finished_at = finished_at1;
+        },
+      Finished
+        {
+          queued_at = queued_at2;
+          started_at = Some started_at2;
+          finished_at = finished_at2;
+        } ) ->
+      cmp_floats queued_at1 queued_at2
+      && cmp_floats started_at1 started_at2
+      && cmp_floats finished_at1 finished_at2
+  | Queued _, (Running _ | Finished _)
+  | Running _, (Queued _ | Finished _)
+  | ( Finished { started_at = None; _ },
+      (Queued _ | Running _ | Finished { started_at = Some _; _ }) )
+  | ( Finished { started_at = Some _; _ },
+      (Queued _ | Running _ | Finished { started_at = None; _ }) ) ->
+      false
+
+let run_times_from_timestamps ~build_created_at
+    ?(current_time = Unix.gettimeofday ()) (ts : timestamps) : run_time_info =
+  match ts with
+  | Queued v -> Queued_for (current_time -. v)
+  | Running { queued_at; started_at } ->
+      Running
+        {
+          queued_for = started_at -. queued_at;
+          ran_for = current_time -. started_at;
+        }
+  | Finished { queued_at; started_at; finished_at } -> (
+      if finished_at < build_created_at then Cached
+      else
+        match started_at with
+        | None ->
+            Finished { queued_for = finished_at -. queued_at; ran_for = None }
+        | Some v ->
+            Finished
+              { queued_for = v -. queued_at; ran_for = Some (finished_at -. v) }
+      )
+
+let timestamps_from_job_info (ji : Client.job_info) :
+    (timestamps, string) result =
+  match (ji.outcome : Client.State.t) with
+  | NotStarted ->
+      Option.fold ji.queued_at
+        ~none:(Error "Queued_at is None - cannot construct timestamp.")
+        ~some:(fun v -> Ok (Queued v))
+  | Active ->
+      Option.fold ji.queued_at
```

Comment:
fwiw I am not happy about this either but it is because `Client.job_info` is populated via a capnp API call. In this implementation I could not get `queued_at` to be just a `float`: 

See https://github.com/ocurrent/ocaml-ci/pull/522/files#diff-cea86c9ace189571331fc2cb600c0636162c5a4e7c32bc8d7377d67333a630ecR230 and thus https://github.com/ocurrent/ocaml-ci/pull/522/files#diff-c45befd13f73a35b60ccfd2a34324712f492068238eafb658d0219d38cc3795bR24

Reply:
That's fair.